clang/lib/AST/ByteCode/Descriptor.cpp

```diff
@@ -468,10 +501,255 @@ bool Descriptor::hasTrivialDtor() const {
 
 bool Descriptor::isUnion() const { return isRecord() && ElemRecord->isUnion(); }
 
+using BlockMoveFn = void (*)(Block *Storage, std::byte *SrcFieldPtr,
+                             std::byte *DstFieldPtr,
+                             const Descriptor *FieldDesc,
+                             const Descriptor *NewDesc);
+
+template <typename T>
+static void moveTy(Block *, std::byte *Src, std::byte *Dst, const Descriptor *,
+                   const Descriptor *) {
+
+  auto *SrcPtr = reinterpret_cast<T *>(Src);
+  if constexpr (!std::is_same_v<T, Floating> &&
+                !std::is_same_v<T, MemberPointer> &&
+                !std::is_same_v<T, Pointer>) {
+  }
+  auto *DstPtr = reinterpret_cast<T *>(Dst);
+  new (DstPtr) T(std::move(*SrcPtr));
+}
+static BlockMoveFn getMovePrim(PrimType Type) {
+  TYPE_SWITCH(Type, return moveTy<T>);
+  llvm_unreachable("unknown PrimType");
+}
+
+/// Moving a primitive array.
+template <typename T>
+static void moveArrayTy(Block *, std::byte *Src, std::byte *Dst,
+                        const Descriptor *D, const Descriptor *NewDesc) {
+  InitMapPtr &SrcIMP = *reinterpret_cast<InitMapPtr *>(Src);
+  SrcIMP = std::nullopt;
+
+  auto *NewInitMap = new (Dst) InitMapPtr(
+      std::make_pair(false, std::make_shared<InitMap>(NewDesc->getNumElems())));
+
+  Src += sizeof(InitMapPtr);
+  Dst += sizeof(InitMapPtr);
+
+  for (unsigned I = 0, NE = D->getNumElemsWithoutFiller(); I < NE; ++I) {
+    auto *SrcPtr = &reinterpret_cast<T *>(Src)[I];
+    auto *DstPtr = &reinterpret_cast<T *>(Dst)[I];
+    new (DstPtr) T(std::move(*SrcPtr));
+    NewInitMap->value().second->initializeElement(I);
+  }
+}
+
+static BlockMoveFn getMoveArrayPrim(PrimType Type) {
+  TYPE_SWITCH(Type, return moveArrayTy<T>);
+  llvm_unreachable("unknown Expr");
+}
+
+static void moveRecord(Block *B, std::byte *Src, std::byte *Dst,
+                       const Descriptor *D, const Descriptor *NewDesc);
+static void moveArrayDesc(Block *B, std::byte *Src, std::byte *Dst,
+                          const Descriptor *D, const Descriptor *NewDesc);
+static BlockMoveFn getMoveFn(const Descriptor *D) {
+  if (D->isPrimitive())
+    return getMovePrim(D->getPrimType());
+  if (D->isPrimitiveArray())
+    return getMoveArrayPrim(D->getPrimType());
+  if (D->isCompositeArray())
+    return moveArrayDesc;
+  if (D->isRecord())
+    return moveRecord;
+
+  return nullptr;
+}
+
+static BlockMoveFn getElemMoveFn(const Descriptor *D) {
+  assert(D->isArray());
+  if (D->isPrimitiveArray())
+    return getMovePrim(D->getPrimType());
+  assert(D->ElemDesc);
+  return getMoveFn(D->ElemDesc);
+}
+
+static void moveRecord(Block *B, std::byte *Src, std::byte *Dst,
+                       const Descriptor *D, const Descriptor *) {
+  assert(D);
+  assert(D->ElemRecord);
+
+  for (const auto &F : D->ElemRecord->fields()) {
+    auto FieldOffset = F.Offset;
+    const auto *SrcDesc =
+        reinterpret_cast<const InlineDescriptor *>(Src + FieldOffset) - 1;
```

Comment:
This feels like it should be a helper someplace, your just repeating the same conjuration all over.